### DISCUSSION TITLE
AdminControlled

### DISCUSSION THREAD

**Discussion by cenwadike** [aurora-is-near/rainbow-bridge#690] [Q&A] [answered]
I'm getting started with rainbow bridge and didn't really understand adminSstoreWithMask function does. Can anyone describe what the inline assembly is supposed to do?

Thanks.

**Comment by mfornet** [comment, marked as answer]
Reference to [the function](https://github.com/aurora-is-near/rainbow-bridge/blob/master/contracts/eth/nearbridge/contracts/AdminControlled.sol#L33-L42).

This functions replace specific bits of the value from the storage at specified key. Only bits that are set in `mask` will be replaced with bits at the same position from `value`.

The motivation for this function is that sometimes, different values are packed inside the same slot on ethereum, either by some solidity optimisation, or our own optimisation. In this case, we only want to replace some part of the value, instead of the whole value. In practice most of the time [adminSstore](https://github.com/aurora-is-near/rainbow-bridge/blob/master/contracts/eth/nearbridge/contracts/AdminControlled.sol#L27-L31) is enough.

How does adminSstoreWork:

```
merged = old_value ^ value
filtered = merged & mask
recovered = filtered ^ old_value
```

1. In the first step (`merged`) we merge the two values together, we can recover either of those by taking the xor with the other one.
2. In the second step we filter all bits, and left only those that are on in the mask. Other bits become `0`
3. In the third step we recover the bits from `value` that were filtered, and the other bits take the value from `old_value` (they where 0 in second step).

Hope this helps to get the idea what is happening under the hood.
